contracts/governance/Checkpoints.sol

```diff
@@ -0,0 +1,267 @@
+    /// @dev Block number must be a finalized block or else this function will
+    ///      revert to prevent misinformation.
+    /// @param account The address of the account to check
+    /// @param blockNumber The block number to get the vote balance at
+    /// @return The number of votes the account had as of the given block
+    function getPastVotes(address account, uint256 blockNumber)
+        public
+        view
+        returns (uint96)
+    {
+        return lookupCheckpoint(_checkpoints[account], blockNumber);
+    }
+
+    /// @notice Retrieve the `totalSupply` at the end of `blockNumber`.
+    ///         Note, this value is the sum of all balances, but it is NOT the
+    ///         sum of all the delegated votes!
+    /// @param blockNumber The block number to get the total supply at
+    /// @dev `blockNumber` must have been already mined
+    function getPastTotalSupply(uint256 blockNumber)
+        public
+        view
+        returns (uint96)
+    {
+        return lookupCheckpoint(_totalSupplyCheckpoints, blockNumber);
+    }
+
+    /// @notice Change delegation for `delegator` to `delegatee`.
+    function delegate(address delegator, address delegatee) internal virtual;
+
+    /// @notice Moves voting power from one delegate to another
+    /// @param src Address of old delegate
+    /// @param dst Address of new delegate
+    /// @param amount Voting power amount to transfer between delegates
+    function moveVotingPower(
+        address src,
+        address dst,
+        uint256 amount
+    ) internal {
+        if (src != dst && amount > 0) {
+            if (src != address(0)) {
+                (uint256 oldWeight, uint256 newWeight) = writeCheckpoint(
+                    _checkpoints[src],
+                    subtract,
+                    amount
+                );
+                emit DelegateVotesChanged(src, oldWeight, newWeight);
+            }
+
+            if (dst != address(0)) {
+                (uint256 oldWeight, uint256 newWeight) = writeCheckpoint(
+                    _checkpoints[dst],
+                    add,
+                    amount
+                );
+                emit DelegateVotesChanged(dst, oldWeight, newWeight);
+            }
+        }
+    }
+
+    /// @notice Writes a new checkpoint based on operating last stored value
+    ///         with a `delta`. Usually, said operation is the `add` or
+    ///         `subtract` functions from this contract, but more complex
+    ///         functions can be passed as parameters.
+    /// @param ckpts The checkpoints array to use
+    /// @param op The function to apply over the last value and the `delta`
+    /// @param delta Variation with respect to last stored value to be used
+    ///              for new checkpoint
+    function writeCheckpoint(
+        uint128[] storage ckpts,
+        function(uint256, uint256) view returns (uint256) op,
+        uint256 delta
+    ) internal returns (uint256 oldWeight, uint256 newWeight) {
```

Comment:
Can weights also be `uint96`? And op and amounts in other places? or there are issues with that?